### PR TITLE
Add support for publishing via Darc + Promotion Pipeline

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes:  #5302 

This PR patches `post-build.yml` to accept new parameters that let the user tell that publishing should be done via Promotion Pipeline, instead of inlining the post-build stages into the build (as is today).

**Note:** 
- Please review the changes ignoring white-space.
- As you'll see this doesn't include support for running the SDL validation via promotion pipeline. More work/discussion needs to happen before that is implemented. [Follow up issue is here.](https://github.com/dotnet/arcade/issues/5372)
  - Related: https://github.com/dotnet/arcade/issues/5389
- Nothing special for reporting errors. I'm thinking to add parameters to let the user inform GitHub aliases that should be pinged when publishing via promotion pipeline fails. That needs more work/discussion. [Follow up issue is here.](https://github.com/dotnet/arcade/issues/5390) 
  - Related: https://github.com/dotnet/arcade/issues/5339

**Tests:**

- [Outlining the stages + not waiting publishing to finish](https://dnceng.visualstudio.com/internal/_build/results?buildId=626196&view=results) 
- [Outlining the stages + waiting publishing to finish.](https://dnceng.visualstudio.com/internal/_build/results?buildId=626167&view=results)
- [Inlining the stages (as is today)](https://dnceng.visualstudio.com/internal/_build/results?buildId=626117&view=results)

### CODE REVIEW

**Comment by JohnTortugo** [comment]
Ping

File: eng/common/post-build/publish-using-darc.ps1 [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    $optionalParams.Add($SigningValidationAdditionalParameters) | Out-Null
+  }
+
+  if ("" -ne $ArtifactsPublishingAdditionalParameters) {
+    $optionalParams.Add("artifact-publishing-parameters") | Out-Null
+    $optionalParams.Add($ArtifactsPublishingAdditionalParameters) | Out-Null
+  }
+
+  if ("false" -eq $WaitPublishingFinish) {
+    $optionalParams.Add("--no-wait") | Out-Null
```
Comment (riarenas) [inline eng/common/post-build/publish-using-darc.ps1:32]: Is the true really needed? thought this was a switch type param.

File: eng/common/post-build/publish-using-darc.ps1 [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    $optionalParams.Add("true") | Out-Null
+  }
+
+  & darc add-build-to-channel `
+	--id $buildId `
```
Comment (riarenas) [inline eng/common/post-build/publish-using-darc.ps1:37]: Does this work for github repos without a GitHubPAT?

File: eng/common/templates/post-build/post-build.yml [hunk 3]
```diff
@@ -44,13 +53,45 @@ parameters:
...
-  dependsOn: ${{ parameters.validateDependsOn }}
-  displayName: Validate
-  variables:
-    - template: common-variables.yml
-  jobs:
-  - template: setup-maestro-vars.yml
+- ${{ if ne(parameters.inline, 'true') }}:
+  - stage: publish_using_darc
+    dependsOn: ${{ parameters.validateDependsOn }}
+    displayName: Publishing using Darc
```
Comment (riarenas) [inline eng/common/templates/post-build/post-build.yml:59]:
```suggestion
    displayName: Publish using Darc
```

File: eng/common/post-build/publish-using-darc.ps1 (hunk 1 above)
Comment (JohnTortugo) [reply eng/common/post-build/publish-using-darc.ps1:32]: It works with the 'true', but I agree that doesn't seem necessary. I'll test/remove it.

File: eng/common/post-build/publish-using-darc.ps1 (hunk 2 above)
Comment (JohnTortugo) [reply eng/common/post-build/publish-using-darc.ps1:37]: Yeah, it was working. I'll triple check this though.

**Comment by JohnTortugo** [comment]
I'm doing a new round of tests for this. Here are some test builds:

Promoting Arcade build: https://dnceng.visualstudio.com/internal/_build/results?buildId=645201&view=results

- Runtime: https://dnceng.visualstudio.com/internal/_build/results?buildId=645388&view=results
- AspnetCore: https://dnceng.visualstudio.com/internal/_build/results?buildId=645390&view=results
- Interactive: https://dnceng.visualstudio.com/internal/_build/results?buildId=645391&view=results
- Winforms: https://dnceng.visualstudio.com/internal/_build/results?buildId=645392&view=results
- Core-SDK/Installer: https://dev.azure.com/dnceng/internal/_build/results?buildId=645395&view=results
- Roslyn: https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3729430&view=results

**Comment by JohnTortugo** [comment]
I'll merge this as my tests passed, except for unrelated errors:

- interactive: https://dnceng.visualstudio.com/internal/_build/results?buildId=650051&view=results
- installer: https://dev.azure.com/dnceng/internal/_build/results?buildId=647063&view=results

- My [tests are failing in Roslyn](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3729849&view=results) because, due to a bug in Azdo.., I'm not able right now to add a secret to a variable group.
- Tests are failing in AspNetCore due to timeouts.. the job shows as [idle for a few hours](https://dnceng.visualstudio.com/internal/_build/results?buildId=649930&view=results) and then fails due to some weird error connecting to an agent.

**Pull request merged**